src/types.ts

```diff
@@ -0,0 +1,80 @@
+export type Merger = <T extends Input, K extends keyof T>(
+  obj: T,
+  key: keyof T,
+  value: T[K]
+) => any;
+
+export type DefuFn = <Source extends Input, Defaults extends Input>(
+  source: Source,
+  ...defaults: Defaults[]
+) => MergeObjects<Source, Defaults>;
+
+export interface Defu {
+  <Source extends Input, Defaults extends Input>(source: Source, ...defaults: Defaults[]): MergeObjects<
+    Source,
+    Defaults
+  >;
+  fn: DefuFn;
+  arrayFn: DefuFn;
+  extend(merger?: Merger): DefuFn;
+}
+
+type Input = Record<string | number | symbol, any>
+
+type MergeArrays<Destination, Source> = Destination extends Array<infer DestinationType>
+  ? Source extends Array<infer SourceType>
+    ? Array<DestinationType | SourceType>
+    : Source | Array<DestinationType>
+  : Source | Destination
+
+type MergeObjects<
+  Destination extends Input,
+  Defaults extends Input
+> = Omit<Destination, keyof Destination & keyof Defaults> & Omit<Defaults, keyof Destination & keyof Defaults> &
+  {
+    -readonly [Key in keyof Destination & keyof Defaults]:
+      Destination[Key] extends null
+        ? Defaults[Key] extends null
+          ? null
+          : Defaults[Key]
+        : Defaults[Key] extends null
+          ? Destination[Key]
+          : Merge<Destination[Key], Defaults[Key]>
+  }
+
+export type Merge<
+  Destination extends Input,
+  Defaults extends Input
+> =
+  // Remove explicitly null types
+  Destination extends null
+  ? Defaults extends null
+    ? null
+    : Defaults
+  : Defaults extends null
+    ? Destination
+    // Handle arrays
+    : Destination extends Array<any>
+      ? Defaults extends Array<any>
+        ? MergeArrays<Destination, Defaults>
+        : Destination | Defaults
+      // Don't attempt to merge Functions, RegExps, Promises
+      : Destination extends Function
```

Comment:
Can't we refactor all types to a union type?

Reply:
Some of these are treated as extending Records so if we're not this verbose we get:
![Screenshot from 2020-10-19 12-45-34](https://user-images.githubusercontent.com/28706372/96446204-00a8cc00-1209-11eb-92c7-5b3a2844f9f7.png)


instead of:
![Screenshot from 2020-10-19 12-44-53](https://user-images.githubusercontent.com/28706372/96446143-ea027500-1208-11eb-8c26-9240b4c5adfe.png)